File: sdk/typescript/src/types/apiKeys.ts
import type { Timestamp, UUID } from "./common.js";
import type { ListResponse } from "./pagination.js";

/**
 * Capability delegated to an API key.
 *
 * - `workspaces:read`: discover and inspect permitted workspaces.
 * - `workspaces:write`: modify permitted existing workspaces; also satisfies `workspaces:read`.
 * - `objects:read`: read, list, and search objects.
 * - `objects:write`: mutate objects; also satisfies `objects:read`.
 * - `attachments:read`: read attachment metadata and content.
 * - `attachments:write`: upload and reuse attachments; also satisfies `attachments:read`.
 * - `graph:read`: read graphs, backlinks, and edges.
 * - `graph:write`: create and revoke edges; also satisfies `graph:read`.
 * - `events:read`: read workspace and object activity.
 * - `realtime:read`: receive ephemeral realtime invalidations for resources the key may read.
 * - `access:manage`: manage workspace memberships, workspace-group links, and object grants.
 * - `admin`: access API-key-enabled global operations, bounded by the owner's current authority.
 */
export type ApiKeyScope =
  | "workspaces:read"
  | "workspaces:write"
  | "objects:read"
  | "objects:write"
  | "attachments:read"
  | "attachments:write"
  | "graph:read"
  | "graph:write"
  | "events:read"
  | "realtime:read"
  | "access:manage"
  | "admin";

/** API key metadata. The secret token is never included after creation. */
export type ApiKey = {
  /** API key ID. */
  id: UUID;
  /** User whose authority the key delegates. */
  user_id: UUID;
  /** Stable user-defined label identifying the key and recorded in audit events. */
  label: string;
  /** Mutable authorization revision. */
  authorization_revision: number;
  /** Capabilities delegated to the key. */
  scopes: ApiKeyScope[];
  /** Workspaces in which the key may exercise workspace-scoped capabilities. */
  workspace_ids: UUID[];
  /** Creation timestamp. */
  created_at: Timestamp;
  /** Last update timestamp. */
  updated_at: Timestamp;
  /** Optional expiration timestamp. */
  expires_at: Timestamp | null;
  /** Revocation timestamp. */
  revoked_at: Timestamp | null;
  /** Last authenticated use timestamp. Updates are intentionally coalesced. */
  last_used_at: Timestamp | null;
};

/** Request body for creating an API key. */
export type CreateApiKeyRequest = {
  /** Stable user-defined label identifying the key and recorded in audit events. */
  label: string;
  /** Capabilities delegated to the key; these can only reduce the owner's authority. */
  scopes: ApiKeyScope[];
  /** Permitted workspaces. An empty list grants no workspace-scoped access. */
  workspace_ids: UUID[];
  /** Optional expiration timestamp. */
  expires_at?: Timestamp | null;
};

/** Request body for replacing an active API key's delegated authorization. */
export type UpdateApiKeyRequest = {
  /** Expected mutable authorization revision. */
  authorization_revision: number;
  /** Replacement delegated capabilities. */
  scopes: ApiKeyScope[];
  /** Replacement permitted workspaces. */
  workspace_ids: UUID[];
};

/** API key response envelope. */
export type ApiKeyResponse = { api_key: ApiKey };

/** API key creation response. The plaintext token is returned exactly once. */
export type CreateApiKeyResponse = ApiKeyResponse & { token: string };

/** API key list response envelope. */
export type ApiKeyListResponse = ListResponse<ApiKey>;

File: sdk/rust/src/wire/api_keys.rs
//! API key resources and requests.

use chrono::{DateTime, Utc};
pub use kival_types::ApiKeyScope;
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
use uuid::Uuid;

use crate::ListResponse;

/// Request body for creating an API key.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
#[serde(deny_unknown_fields)]
pub struct CreateApiKeyRequest {
    /// Stable user-defined label identifying the key and recorded in audit events.
    pub label: String,
    /// Capabilities delegated to the key. These can only reduce the owning user's authority.
    pub scopes: Vec<ApiKeyScope>,
    /// Workspaces in which the key may exercise workspace-scoped capabilities.
    /// An empty list grants no workspace-scoped access.
    pub workspace_ids: Vec<Uuid>,
    /// Optional expiration timestamp.
    pub expires_at: Option<DateTime<Utc>>,
}

/// Request body for replacing an active API key's delegated authorization.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
#[serde(deny_unknown_fields)]
pub struct UpdateApiKeyRequest {
    /// Expected mutable authorization revision.
    pub authorization_revision: i32,

    /// Capabilities delegated to the key. These can only reduce the owning user's authority.
    pub scopes: Vec<ApiKeyScope>,
    /// Workspaces in which the key may exercise workspace-scoped capabilities.
    /// An empty list grants no workspace-scoped access.
    pub workspace_ids: Vec<Uuid>,
}

/// API key metadata. The secret token is never included after creation.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
pub struct ApiKey {
    /// API key ID.
    pub id: Uuid,
    /// User whose authority the key delegates.
    pub user_id: Uuid,
    /// Stable user-defined label identifying the key and recorded in audit events.
    pub label: String,
    /// Mutable authorization revision.
    pub authorization_revision: i32,
    /// Capabilities delegated to the key.
    pub scopes: Vec<ApiKeyScope>,
    /// Workspaces in which the key may exercise workspace-scoped capabilities.
    pub workspace_ids: Vec<Uuid>,
    /// Creation timestamp.
    pub created_at: DateTime<Utc>,
    /// Last update timestamp.
    pub updated_at: DateTime<Utc>,
    /// Optional expiration timestamp.
    pub expires_at: Option<DateTime<Utc>>,
    /// Revocation timestamp.
    pub revoked_at: Option<DateTime<Utc>>,
    /// Last authenticated use timestamp. Updates are intentionally coalesced.
    pub last_used_at: Option<DateTime<Utc>>,
}

/// API key response envelope.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
pub struct ApiKeyResponse {
    /// API key metadata.
    pub api_key: ApiKey,
}

/// API key creation response. The token is returned exactly once.
#[derive(Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
pub struct CreateApiKeyResponse {
    /// Created API key metadata.
    pub api_key: ApiKey,
    /// Plaintext bearer token. Kival does not store this value.
    pub token: String,
}

/// API key list response envelope.
pub type ApiKeyListResponse = ListResponse<ApiKey>;

impl std::fmt::Debug for CreateApiKeyResponse {
    fn fmt(&self, formatter: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        formatter
            .debug_struct("CreateApiKeyResponse")
            .field("api_key", &self.api_key)
            .field("token", &"[REDACTED]")
            .finish()
    }
}

#[cfg(test)]
mod tests {
    use crate::ApiKeyScope;

    #[test]
    fn all_scopes_round_trip_through_stable_wire_names() {
        for scope in ApiKeyScope::ALL {
            assert_eq!(scope.as_str().parse(), Ok(*scope));

            let json = serde_json::to_string(scope).expect("scope should serialize");
            let decoded =
                serde_json::from_str::<ApiKeyScope>(&json).expect("scope should deserialize");
            assert_eq!(decoded, *scope);
        }
    }
}
